torch/csrc/jit/passes/onnx/peephole.cpp

```diff
@@ -167,6 +172,66 @@ void fuseTransposeIntoGemm(std::shared_ptr<Graph>& graph) {
   }
 }
 
+void pushPackingPastRnn(std::shared_ptr<Graph>& graph) {
+  for (auto it = graph->nodes().begin(); it != graph->nodes().end(); ++it) {
+    auto* n = *it;
+
+    if (n->kind() != kPackPadded) {
+      continue;
+    }
+    if (n->outputs()[0]->uses().size() != 1) {
+      // For now, only handle the case where there is one consumer.
+      continue;
+    }
+    Node * rnn = n->outputs()[0]->uses()[0].user;
+    if (!isRNN(rnn)) {
+      continue;
+    }
+
+    // remove PackPadded from in front of the RNN
+    n->outputs()[0]->replaceFirstUseWith(n->inputs()[0]);
```

Comment:
Minor style point (you don't have to fix): you can access a particular input with `n->input(0)` and `n->output(0)`.

Reply:
++

Reply:
If there's only one use of `n->outputs()[0]`, you didn't need to add a `replaceFirstUseWith` function, did you?

I can't think of many situations where you'd actually want `replaceFirstUseWith`.

Reply:
(also answering the next comment here)

there is only one use of `n->outputs()[0]` (enforced by line 198), so yes I will switch this to replaceAllUsesWith.

there are actually multiple consumers of `n->outputs()[1]`, and this is fine. The same blob feeds into each level of the RNN as the sequence_lens input (see the code `sequence_lens = ...` in symbolic.py).

Note that pytorch RNNs do a weird thing - at each level, they both consume and return the lengths/batch_sizes (as the second element of the PackedSequence tuple), even though the returned lengths must be the same as the input lengths. ONNX RNNs don't do that weird thing, they only consume the lengths. Therefore I can't mimic the pytorch weirdness exactly, and instead must share the lengths blob between layers